boards/shared/Src/Io/Io_SharedAdc.c

```diff
@@ -1,169 +1,140 @@
+    if (alloc_index == 0U)
     {
-        status = false;
-        Error_Handler();
+        // Initialize ADC instance containing VREFINT before initializing other
+        // ADCs
+        assert(vrefint_index <= adc_input->num_active_channels);
+
+        adc_inputs->vrefint_index = vrefint_index - 1;
+        // Store the pointer pointing to the measured VREFINT value
+        measured_vrefint_address =
+            &adc_input->adc_readings[adc_input->vrefint_index];
     }
-
-    // Only set adc_initialized to true if all initialization steps are
-    // error-free
-    if (status == true)
+    else
     {
-        adc_initialized = true;
+        UNUSED(vrefint_index);
     }
-}
-
-uint32_t Io_SharedAdc_GetAdcMaxValue(void)
-{
-    CheckAdcInitialized();
-
-    return (const uint32_t)adc_max_value;
-}
-
-const uint32_t *Io_SharedAdc_GetAdcValues(void)
-{
-    CheckAdcInitialized();
 
-    return (const uint32_t *)adc_values;
+    alloc_index++;
+    return adc_input;
 }
 
-float32_t Io_SharedAdc_GetActualVdda(void)
+float Io_SharedAdc_GetChannelVoltage(
+    const struct AdcInput *const adc_input,
+    size_t                       adc_rank)
 {
-    CheckAdcInitialized();
+    assert(adc_rank > 0 && adc_rank <= adc_input->num_active_channels);
 
-    // The actual VDDA voltage supplying the microcontroller can be calculated
-    // using the following formula:
+    // This function performs the following:
+    // (1) Calculate the microcontroller's actual VDDA value using the following
+    // formula
     //
-    //         3.3V x VREFINT_CAL
-    // VDDA = ---------------------
-    //           VREFINT_DATA
+    //                    3.3V x VREFINT_CAL
+    //  ACTUAL_VDDA = ---------------------------
+    //                      VREFINT_MEASURED
     //
-    // Where:
-    // - VREFINT_CAL is the VREFINT calibration value
-    // - VREFINT_DATA is the actual VREFINT output value converted by the ADC
+    // VREFINT_CAL is the VREFINT calibration value (stored at 0x1FFF7BA)
+    // VREFINT_MEASURED is the actual VREFINT output value converted by an ADC
+    //
+    // (2) Calculate the voltage of the given ADC Channel
+    //
+    //                 ACTUAL_VDDA x ADCx_READINGS
+    //  V_CHANNELx = ----------------------------
+    //                       ADC_MAX_VALUE
+    //
+    // ACTUAL_VDDA is the vdda value computed in step (1)
+    // ADCx_DATA is the value of the channel measured by the ADC
+    // ADC_MAX_VALUE is the maximum ADC value
 
-    return (float32_t)(
-        3.3f * (float32_t)(GetVrefintCalibrationValue()) /
-        (float)adc_values[vrefint_index]);
+    const uint16_t *const VREFINT_CALIBRATION_ADDRESS =
+        (uint16_t *)(0x1FFFF7BA);
+    const float ACTUAL_VDDA =
+        (float)(3.3f * (uint16_t)(*VREFINT_CALIBRATION_ADDRESS) / *measured_vrefint_address);
+
+    return (
+        float)(ACTUAL_VDDA * adc_input->adc_readings[adc_rank - 1] / adc_input->adc_max_value);
```

Comment:
Could we break this out a bit? Something like:
```
        float A = (float)ACTUAL_VDDA * adc_input->adc_readings[adc_rank - 1];
        float B = A / (float)adc_input->adc_max_value;
        return B;
```
Where `A` and `B` are replaced with more descriptive names (_we don't need to split it up exactly like this, just trying to understand what we're calculating here_).